### ISSUE TITLE
Flutter should support iOS 15's Secure Paste feature

### ISSUE THREAD

**Issue by justinmc** [flutter/flutter#103163]
iOS 14 introduced a system notification when an app pasted from the pasteboard as a security feature.  In iOS 15, Apple is introducing "secure paste", which seems like it will bypass this notification when the paste is initiated by the user.

https://www.apple.com/newsroom/2021/06/apple-advances-its-privacy-leadership-with-ios-15-ipados-15-macos-monterey-and-watchos-8/
https://developer.apple.com/videos/play/wwdc2021/10085/?time=360

That information about the feature is vague, but it may not be possible for Flutter to implement this, at least for the text selection menu, since Flutter doesn't use the native text selection menu.  If I'm understanding it correctly, tapping on Flutter's "paste" button will always be treated as being initiated by the app, and it will show the notification.

We may be able to support this for the keyboard by receiving whatever "user paste" event Apple creates and sending something along to the framework.

CC @jmagman

**Comment by jmagman** [comment]
The WWDC talk calls it "secure paste" and I _think_ that's the same thing as "paste and go" in the docs https://developer.apple.com/documentation/uikit/uiactionpasteandgo

**Comment by jmagman** [comment]
See `UIResponderStandardEditActions` instance method [`pasteAndGo:`](https://developer.apple.com/documentation/uikit/uiresponderstandardeditactions/3750907-pasteandgo?language=objc) [`pasteAndMatchStyle:`](https://developer.apple.com/documentation/uikit/uiresponderstandardeditactions/3750908-pasteandmatchstyle?language=objc) and [`pasteAndSearch:`](https://developer.apple.com/documentation/uikit/uiresponderstandardeditactions/3750909-pasteandsearch?language=objc)

Relevant engine code:
https://github.com/flutter/engine/blob/9d784a8b0f77a40829ee38e5914cc9154f241bfb/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm#L1102

**Comment by jmagman** [comment]
Mentioned here as well https://developer.apple.com/videos/play/wwdc2021/10059/?time=1421 there's a serious lack of documentation about this...

**Comment by hellohuanlin** [comment]
wow my UIKit knowledge is seriously outdated.

**Comment by jmagman** [comment]
> The WWDC talk calls it "secure paste" and I _think_ that's the same thing as "paste and go" in the docs https://developer.apple.com/documentation/uikit/uiactionpasteandgo

I think was totally off and they aren't the same, I see this comment in https://developer.apple.com/forums/thread/698209 from "Frameworks Engineer "

> pasteAndGo is not implemented on UITextField because the text field doesn't have any action associated. These methods are meant to paste and then do something, e.g. in Safari where you can paste and then immediately load the URL you pasted.

Okay I'm not sure how to implement this, maybe we can't since Flutter is always grabbing from the pasteboard programmatically.

**Comment by mazzonem** [comment]
Any update on this, iOS 16 is now showing dialog every time we paste to a Textfield.

Will flutter be able to avoid grabbing programmatically?

**Comment by jmagman** [comment]
> Any update on this, iOS 16 is now showing dialog every time we paste to a Textfield.

You mean the behavior is different from iOS 15?

**Comment by limenote135** [comment]
Is there any idea to do this?
Only shows the notification on iOS15 but shows the prompt every time on iOS16 when we paste some texts from the clipboard into TextField. Even if we tap "Allow paste", the prompt shows again if the texts of the clipboard are changed.

**Comment by orestesgaolin** [comment]
Definitely one of the top irritating things right now on iOS with Flutter. Some users unfortunately don't understand the technical reasons behind the popup. For instance our user can paste the OTP code that they receive via e-mail. We recently got 1 star just because of the clipboard popup:

> On very first open, app tries to access what is in my clipboard. This is a sure sign to me that the app is not designed with privacy in mind. Hope you can revisit this, developers. God knows we need an app like this.

**Comment by acalatrava** [comment]
Is there any workaround on this? Is this technically possible to be fixed?

**Comment by justinmc** [comment]
iOS 16.1 recently added an app setting called "Paste from other apps".  This is my update on the current state of this issue based on some exploration I did.

## Currently with Flutter
I ran a Flutter app on an iOS 16.1 simulator with just a simple TextField.  This is what I saw when changing "Paste from other apps" to each of its three values in settings for my Flutter app.

| Paste permission dialog | Paste notification |
| --- | --- |
| <img width="492" alt="Screenshot 2023-01-03 at 10 26 05 AM" src="https://user-images.githubusercontent.com/389558/210418571-7b54e3c5-9178-45d8-9ec7-9416ec27b4bb.png"> | <img width="319" alt="Screenshot 2023-01-03 at 10 30 26 AM" src="https://user-images.githubusercontent.com/389558/210419185-b57d30a3-a70e-46b6-a5be-e90136273f5a.png"> |

  * Ask (default) - Shows a dialog (screenshot above) where the user can choose "allow" or "don't allow".  If "don't allow" is chosen, then nothing is pasted and the context menu remains open.  If "allow" is chosen, the text is pasted and the context menu is closed like normal.  Pasting the same text again will not show the dialog again.  Pasting any other text, including new text from the same app ("Reminders" in the case of the screenshot), will show the dialog again.  The paste notification (screenshot above) is never shown.
  * Allow - Always allows paste from any app.  The paste notification is always shown on paste but the dialog is never shown.
  * Deny - Nothing happens when the paste button is pressed.  The context menu is not closed.

## Currently with native apps

I tried this by running a SwiftUI app on the same iOS 16.1 simulator, but I wonder if it behaves differently for production apps.  Can anyone verify what happens on native apps?

Based on what I saw, pasting with the native text selection toolbar does not trigger the permission dialog or the paste notification.  Only if the app programmatically accessed the clipboard would the dialog, the notification, and the "Paste from other apps" setting matter.

## What can Flutter do about it?

I don't see any clear action that Flutter can take to mitigate this!  @sudha-anecure, you asked for a dialog to change this setting in https://github.com/flutter/flutter/issues/116599.  I'm not aware of a way to programmatically change this setting, are you?  We could send users to the app settings with an explanation of what to do, but I'm not sure that's something we want to build into the framework.

Any other ideas?  Currently this is clearly an inferior experience to native.

**Comment by justinmc** [comment]
@orestesgaolin How can I reproduce the clipboard access "on very first open" that you posted about above?

**Comment by AugustoResende** [comment]
It would be nice if there was a way to simulate a native input that opens the native "paste menu" that doesn't generate the popup

**Comment by justinmc** [comment]
There are some packages that allow you to use native text fields in Flutter, with their own pros and cons.  e.g. https://pub.dev/packages/flutter_native_text_input

**Comment by n13** [comment]
Currently on iOS, native developers get around this annoyance by having a clearly marked special paste button

UIPasteControl

https://developer.apple.com/documentation/uikit/uipastecontrol

This is a button like any other button, but it needs to be marked "Paste" (it's a little unclear what the requirements are or who is checking with the requirements?! - maybe app review?). 

Then, when the app is using this button, there's no special paste dialog. User can just paste. 

We need this for onboarding, since the dynamic links use the pasteboard to uniquely identify a device / link payload. 

I am not sure if this is feasible to implement in Flutter.

Other exceptions to the annoying dialog are user explicitly using UITextField controls to paste. So putting a native text field somewhere might work for some cases (not for our use case).

**Comment by justinmc** [comment]
@n13 Thanks for the analysis, I wasn't aware of UIPasteControl.  It wouldn't be easy to show this native paste button in the context menu, but it might be the only somewhat feasible option for Flutter to pursue here.

For your onboarding paste idea, I was under the impression that iOS would prompt you again each time you copy something different.  That was what I experienced on iOS 16.1 (see my [comment above](https://github.com/flutter/flutter/issues/103163#issuecomment-1370120640)).  But if that has changed let me know, it wouldn't be so bad if it only appeared once.

**Comment by alxmoroz** [comment]
@justinmc @n13 
Seems like UIPasteControl implemented here: https://pub.dev/packages/flutter_native_text_input
Why is the Flutter team unable to do the same?

**Comment by taufiq-husain** [comment]
This only happens in Flutter apps on iOS. On one of my apps, users can share OR paste links only. So I have found an alternative workaround where sharing links to my app is an intent from browsers and such. Now when you're on a browser and tap the share button, my app shows up as an option. Just a suggestion for people needing paste features from other apps often.

Either way, this is a much-needed fix. I don't understand why paste from Apple's native action menu shows the popup though?

**Comment by shenghan97** [comment]
> I don't understand why paste from Apple's native action menu shows the popup though? It doesn't do this on any other apps I use currently?

If you're referring to the edit menu that pops up when you tap or long press in the text field, it's because  text selection edit menu in flutter isn't really "native" (unless you use plugins like `flutter_native_text_input `). Flutter only mimics the the UI of the native menu, but renders it completely by Flutter's own graphic engine (Skia as for now). In native apps, the menu is rendered by the OS instead, so it doesn't pop up that permission dialog or the heads up.

**Comment by justinmc** [comment]
That's right.  That's also the answer to @alxmoroz's question above.  flutter_native_text_input doesn't appear to be using UIPasteControl if I'm not mistaken, it's just showing the native context menu and paste button.

**Comment by justinmc** [comment]
I just quickly double checked the behavior on iOS 16.4 to see if anything has changed, but it seems the behavior is still the same as in my [comment above](https://github.com/flutter/flutter/issues/103163#issuecomment-1370120640).

**Comment by taufiq-husain** [comment]
> 
> > I don't understand why paste from Apple's native action menu shows the popup though? It doesn't do this on any other apps I use currently?
> 
> If you're referring to the edit menu that pops up when you tap or long press in the text field, it's because  text selection edit menu in flutter isn't really "native" (unless you use plugins like `flutter_native_text_input `). Flutter only mimics the the UI of the native menu, but renders it completely by Flutter's own graphic engine (Skia as for now). In native apps, the menu is rendered by the OS instead, so it doesn't pop up that permission dialog or the heads up. 

I tried using this but the library hasn't been updated in a while and it's not as customizable as I want for my app. Also get weird UI errors sometimes.

**Comment by shw2ypro** [comment]
Hello
I faced this issue when using firebase dynamic link
the user should allow the prompet to complete the invitation process successfully
How can i solve this issue?

**Comment by jaredanderton** [comment]
@shw2ypro  See step 3 (optional) for bypassing the "paste from" prompt for firebase dynamic links: 
https://firebase.google.com/docs/dynamic-links/ios/receive

**Comment by limenote135** [comment]
I noticed Google Cloud app is made by Flutter but it doesn't show this prompt. How does it implement??

**Comment by shenghan97** [comment]
Are you sure its built with flutter? It doesn't seem like they use the flutter text input widget. 



> I noticed Google Cloud app is made by Flutter but it doesn't show this prompt. How does it implement??

**Comment by rotorgames** [comment]
Hi there. I'm affected by this issue as well. Seems working with input fields in Flutter getting worst over time. 
I know that's a conceptual issue and can't be fixed easily cuz Flutter draws input fields by itself and doesn't use platform's components. Maybe, in over time platforms will apply more restrictions that will make using of input fields even more worst. 

Do we have a chance to start looking at another way working with input fields? Maybe Flutter starts using native input fields behind the scene to support native features including copy, pasting, prevent "past alerts" and etc or even redesign all input fields that will work with native input fields by default?

I know it sounds like a lot of work and deviation of the Flutter's ideology but it seems that we can draw a lot of elements like buttons, images, tab bars, etc, but not input fields cuz there're deep interlinked with the platform functionality, so every day it becomes hardest and hardest to support/simulate native behaviour for input fields. 

Tbh, I like that Flutter draws all input fields by itself cuz we have more flexibility for customising them. So I prefer using a hybrid approach when we continue drawing input fields by Flutter but using a native field to support platform related functionality. But I'm not really sure this is possible at all. 

Thoughts?

**Comment by rotorgames** [comment]
Just FYI. I've been playing with UIButton and UIAction.Identifier.paste. Unfortunately it works only with UIMenu. So, it's not possible to use UIButton for handling "past in" functionality with no system prompting. 

However, I found out that UIPastControl can be integrated into contextMenuBuilder of TextField via UiKitView (PlatformView). Actually we can put UIPastControl behind the CupertinoTextSelectionToolbarButton (or as a background) and make CupertinoTextSelectionToolbarButton input transparent to pass all touch events to UIPastControl. The native class that implements FlutterPlatformView for the control can also handle pasting by implementing UIPasteConfigurationSupporting so we even don't need a UITextView here.

But, I'm not sure if Apple accept that approach during app review although CupertinoTextSelectionToolbarButton does exactly that UIPastControl does but we're hiding the second one so that can be a matter of getting rejected. 

But again, technically it's possible.

**Comment by rotorgames** [comment]
Here's what I've achieved today. 

The right "past" button is a Flutter's CupertinoTextSelectionToolbarButton that uses clipboard directly. So, we get a system prompt.
The left "past" is a UIPasteControl that's integrated to the context menu via PlatformView. As you can see, by clicking the item we don't get the prompts anymore. 

As result, we have a really good chance to integrate copy-past feature on iOS 16 or higher with no system prompting.

P.S. Please, keep in mind that I adjusted size of the UIPastControl to make it look like a Flutter's item. However UIPastControl looks bigger by default. 

https://github.com/flutter/flutter/assets/1790504/98a2c8bf-9f75-4caa-ae3f-91f9540ccbe9

**Comment by hellohuanlin** [comment]
@rotorgames hi. Thanks for the research. Could you share the platform view implementation so I can play around with it? Thanks.

**Comment by rotorgames** [comment]
@hellohuanlin Sure thing. 

Here you can get the project example to play. https://github.com/rotorgames/flutter_textview_iospaste

**Comment by rotorgames** [comment]
Hey @hellohuanlin did you have a chance to take a look of this hight priority (P1) task?

**Comment by hellohuanlin** [comment]
> Hey @hellohuanlin did you have a chance to take a look of this hight priority (P1) task? 

Yeah still on it. Sorry for the delay. Got distracted by multiple iOS 17 urgent bugs.

**Comment by alxmoroz** [comment]
> It's now possible to avoid the paste notification by using the native text selection context menu in your iOS 16+ apps

Hooray! Thank you!
We are really looking forward to it appearing in the stable.

**Comment by nicolasvahidzein** [comment]
Link is not available.

**Comment by AugustoResende** [comment]
@justinmc something has been wrong in the merge apparently

**Comment by hellohuanlin** [comment]
> Link is not available.

The PR was reverted and now has been re-landed. The link is:
 https://github.com/flutter/flutter/blob/8e35cf5125e573e03498c93fa3cacf686d650b4d/examples/api/lib/widgets/system_context_menu/system_context_menu.0.dart

**Comment by alvindrakes** [comment]
I have upgraded to Flutter version 3.22.2, 

Is this change reverted again? 

Can't seem to find `SystemContextMenu`?

**Comment by justinmc** [comment]
@alvindrakes This is not in the stable channel yet. https://justinmc.github.io/flutter_releases/#/pr/framework/148265

**Comment by sfshaza2** [comment]
Hey, gang! (@jmagman @justinmc @sethladd), let me know if this requires doc changes to docs.flutter.dev for the Q3 release. If so, we should talk.

**Comment by hellohuanlin** [comment]
@sfshaza2 Once we turn system context menu on by default, then 

(1) For most common use case (without crazy menu customization), developers should get it out of box by putting a regular `TextField` on screen, without touching `contextMenuBuilder` API, so we don't have to worry about it. 

(2) For complex use case with customization, the API doc will be updated with our code change. Is there any other doc that we need to cover? @justinmc

**Comment by anhquangtech** [comment]
- Update your project flutter to version 3.24.1. (use [fvm](https://fvm.app/))
- Add [SystemContextMenu](https://api.flutter.dev/flutter/widgets/SystemContextMenu-class.html#:~:text=The%20context%20menu%20is%20the,menu%20visible%20at%20a%20time.) to your Text Field
```dart
contextMenuBuilder:
          (BuildContext context, EditableTextState editableTextState) {
        if (SystemContextMenu.isSupported(context)) {
          return SystemContextMenu.editableText(
            editableTextState: editableTextState,
          );
        }
        return AdaptiveTextSelectionToolbar.editableText(
          editableTextState: editableTextState,
        );
      },
```
This works for me

**Comment by yuukiw00w** [comment]
TextField looks good.
But `SearchBar` does not yet have a contextMenuBuilder property.
Any widget that uses a TextField internally, like `SearchBar`, may have the same problem.

**Comment by justinmc** [comment]
@yuukiw00w Good catch! I opened a PR: https://github.com/flutter/flutter/pull/154833

**Comment by mahapsub** [comment]
Any update on this? The app seems to ask for permission upon booting.

**Comment by Tobi14601** [comment]
> * Update your project flutter to version 3.24.1. (use [fvm](https://fvm.app/))
> * Add [SystemContextMenu](https://api.flutter.dev/flutter/widgets/SystemContextMenu-class.html#:~:text=The%20context%20menu%20is%20the,menu%20visible%20at%20a%20time.) to your Text Field
> 
> ```dart
> TextField(
>       // Add this
>       contextMenuBuilder:
>           (BuildContext context, EditableTextState editableTextState) {
>         // If supported, show the system context menu.
>         if (SystemContextMenu.isSupported(context)) {
>           return SystemContextMenu.editableText(
>             editableTextState: editableTextState,
>           );
>         }
>         // Otherwise, show the flutter-rendered context menu for the current
>         // platform.
>         return AdaptiveTextSelectionToolbar.editableText(
>           editableTextState: editableTextState,
>         );
>       },
>     )
> ```
> 
> This works for me.

This also fixed the annoying system alert for me. 

Is it possible to add translations to the SystemContextMenu?

My App supports multiple languages but the shown ContextMenu is always in english even when the phone is set to another language.  Is there a fix / workaround for that?

**Comment by hellohuanlin** [comment]
> Is it possible to add translations to the SystemContextMenu?

@Tobi14601 Both flutter menu and native menu should already been localized. If not, can you file a separate ticket with your code?

**Comment by farrerdev** [comment]
Anyway to custom button items for SystemContextMenu like AdaptiveTextSelectionToolbar.buttonItems?

**Comment by Tobi14601** [comment]
> @Tobi14601 Both flutter menu and native menu should already been localized. If not, can you file a separate ticket with your code?

I added a new language and forgot to add it in the iOS App bundle. Adding it fixed my issue.

**Comment by justinmc** [comment]
@farrerdev That functionality is coming in steps. See https://github.com/flutter/flutter/pull/159013 and https://github.com/flutter/engine/pull/56362.

**Comment by jmagman** [comment]
Good timing, bot.
https://github.com/flutter/flutter/issues/163067 merged today. The "Don't Allow Paste" pop-up should no longer appear on the master branch (assuming no custom context menu actions) when the user pastes into a text field.

**Comment by sethladd** [comment]
yay!

**Comment by farrerdev** [comment]
Can you support using default items like copy/paste but override onPressed for IOSSystemContextMenuItem, like ContextMenuButtonItem? @justinmc

**Comment by hellohuanlin** [comment]
Closing as completed after https://github.com/flutter/flutter/pull/171825 thanks to @jingshao-code 

There's also sub-menus that we don't support yet, but it's also not supported in existing Flutter context menu, so we do have feature parity now.
